bot/cogs/doc/cog.py

```diff
@@ -52,18 +53,83 @@
 NOT_FOUND_DELETE_DELAY = RedirectOutput.delete_delay
 
 
+class TODO_PLACEHOLDER:
+    __slots__ = ("queue", "item_events", "started", "fetch_lock")
+    results_dict: Dict[str, discord.Embed] = None
```

Comment:
This class handles parsing of all items on one page, for example behind the url https://docs.python.org/3/library/random.html it'll parse the module itself, all the classes etc. using the list as a FIFO queue for prioritizing items when a doc request is made.

I'm not quite sure how to handle the results here because `wait_for_item` is an async task that is started on the first request to an url. Currently it is done by setting the above `results_dict` to the dict which the cog fetches results from, but it doesn't feel great. Something with callbacks came to mind  but the interface for the caller wouldn't be much better

Reply:
Yeah, I don't like it either. Yielding/returning the results makes more sense to me. Whatever is using this class should be responsible for storing the results.

Reply:
The goal here was to simplify it a bit for the user class
The only thing that came to mind was starting a task in the cog that'd iterate over `parse_items` (which would yield the descriptions) and assign to its result dict but that doesn't feel like a great interface either as it puts the responsibility of starting the parsing on the class that uses it.

Reply:
Yeah. Then how about making this class act as the cache container itself, or at least storing the cache solely in this class rather than passing a dict around? You can re-design this class a bit so that you're not going to need a separate instance for each page, which will make my suggestion viable. The queue can hold futures for all pages; you'll just need to keep track of which pages have already been added to the queue to avoid redundant parsing.

Reply:
Just so I'm sure we're thinking of the thing.
There'd be a "master" queue where all pages would get queued as requested from an internal `url: symbol_ids`  mapping, the results of that would be saved to something like `self.objects` in the cog. Then the public API of the class could just be an `add_symbol_id(url, symbol_id)` when initiating inventories and `get_symbol(symbol_id)` for requesting them

Reply:
Yes, something along those lines. You could probably have those methods accept a `DocItem`, but if you only need those two attributes then I suppose being more explicit is better. That being said, I thought of two problems:

1. When should the queue start? Currently it runs for only one page once a specific symbol is requested. If the queue is shared, I'm not sure how to still accomplish this. Possibly store the symbols in a separate data structure and move them to the queue later, once one symbol from the page is requested. That's just one idea for a solution.
2. How should the soup objects be stored? How will it know when to get rid of them?

Reply:
Yeah passing the DocItems through is better since we need the url anyway so there's no need for it to be passed by the caller, and they already exist.
Started trying out the implementation, the results and `self.urls` are moved from the Cog to this class. On a item request , if the symbols need to be queued, the `DocItem`s of the page that symbol is located in (from `self.urls`) is moved to the queue and the queue is started. Together witch each symbol we can also pass the soup in a container which takes care of their storage (Once we're done with the symbols from the page it gets completely dereferenced on our part). When a symbol is parsed it's handled like before, but the result is returned with the events and result dict handled internally.

Should the parsing task be stopped when the queue ends up empty again? We can either start it every time symbols are added to an empty queue, or have it run infinitely from the first request and just skip the code when there's nothing to handle

Reply:
That sounds like what I had in mind. I'm happy with that solution. What about you?

> Should the parsing task be stopped when the queue ends up empty again?

Skip how? If it could await until an item was put in the queue, that'd be fine. Unfortunately, you can't use an asyncio queue since you need to be able to move items from arbitrary positions to the front. Speaking of which, is there a more efficient data structure than a list for doing that? Earlier, I was looking into [priority queues](https://stackoverflow.com/questions/2288241/priority-queue-with-dynamic-item-priorities). Anyway, restarting the tasks sounds better, but I don't know for sure until you clarify the implementation.

Reply:
I haven't found a more fitting data structure, tried priority queues but it looks like they only do the priority matching on item insertion, and it's done by comparing it against all other items which also isn't very efficient.

Reply:
Pushed the changed class in c1a516d372f5c6af5ff30964748e4178ee36b238